docs/source/config_settings.rst
.. _ConfigurationSettings:

RadSSH Configuration Settings
=============================

RadSSH uses a number of configuration settings that may be of interest to customize and tweak the runtime behavior of the shell, its plugins, and/or the API Cluster behavior. While the default settings are intended to be generally appropriate for most environments, they are user-configurable to allow tailoring on a per-user or per-environment need.

The basic default settings are embedded in the RadSSH module itself, but can be amended by up to three additional stages of configuration: System Configuration, User Configuration, and Command Line Options.

To view the predefined default settings, or to produce a template file for a System Configuration file or User Configuration file, you can run ``python -m radssh.config``. This utility will output a sample settings file in comment form. If you save the output to a file, you will need to uncomment specific lines (and change values) in order to enable RadSSH configuration changes.

In general:
 - Blank lines are ignored
 - Lines starting with **#** are ignored (comment lines)
 - All other lines should be in the form **keyword=value**, where **keyword** is the option name to set, and **value** is the value to set.

System Level Configuration
--------------------------

If the file ``/etc/radssh_config`` exists, then the settings will be read in from that file. Any option settings found in the System Configuration file will supercede (not add to) the default setting. It is not necessary to specify every configuration option in the System Configuration file; only the settings you have need to change from the default.

**Important** If the System Configuration file sets **user.settings** to a blank string, then the RadSSH shell will not read settings from any User Configuration file, and will also ignore any command line options that would otherwise affect settings. If the **user.settings** option points to a file or pathname, then any settings in that file may override default or system values.

User Level Configuration
------------------------

By default, ``~/.radssh_config`` is used, but this location may be changed (or removed) by an optional System Level Configuration. Any option settings found in the User Configuration file will supercede the setting values from the default configuration or the System Configuration. The sole exception is the **user.settings** option; altering its value will have no effect if RadSSH is already processing the user settings file.

Command Line Options
--------------------

Settings provided on the command line take the highest precedence. Command line options of the form **keyword=value** override settings that are read from the User Configuration file (or any earlier location).


--------


General Settings
----------------
 - username (default: **$SSH_USER** or **$USER**)
    Login name for establishing SSH sessions
 - loglevel (default: ERROR)
    One of CRITICAL, ERROR, WARNING, INFO, DEBUG. The old **verbose** setting is now deprecated.
 - output_mode (default: stream)
    **stream** will output lines of text to the console as they come in. **ordered** will preserve host ordering, which may give the appearance of disrupting parallelism on commands with lengthy output. **grouped** ensures completed output per host, without interleaving lines. **off** turns off console output while commands are running, but does not affect file logging of output. Can be changed within the shell via the **\*output** command.
 - max_threads (default: 120)
    Limit RadSSH processing threads. Independent of the baseline 1 thread per SSH connection overhead.
 - shell.console (default: color)
    Set to **mono** if the output color coding is not desired.
 - shell.prompt (default: "RadSSH $")
    The RadSSH shell prompt issued before reading each command line.
 - logdir (default: session_%Y%m%d_%H%M%S)
    Location of the session logging directory. Defaults to current directory, and supports expansion of datetime elements. Path prefix can be added, including ~ for user home directory. Can be set to blank, to cause logging to go to the console as stderr stream.
 - log_out (default: out.log)
    Consolidated (all host) stdout filename. Created in the logdir directory.
 - log_err (default: err.log)
    Consolidated (all host) stderr filename. Created in the logdir directory.
 - historyfile (default: ~/.radssh_history)
    Save command line history across sessions to this file.
 - socket.timeout (default: 30)
    Network connection and read/write timeout (in seconds).
 - keepalive (default: 180)
    Send periodic network traffic to prevent connections from being terminated due to being idle.
 - hostkey.verify (default: reject)
    Determines how RadSSH handles verification of remote host keys against the ~/.ssh/known_hosts file. **reject** will reject connections if the remote host key is not already validated and accepted in the known hosts file. Other, less secure options include **prompt** which will interactively ask the user to accept unrecognized keys, **accept_new** which will automatically accept new entries, but reject if a previously accepted key no longer matches, and **ignore** which bypasses host key verification completely.
 - hostkey.known_hosts (default: ~/.ssh/known_hosts)
    Change to a new file path to keep RadSSH known hosts completely separate from OpenSSH client known hosts. If they share the common default file, keys accepted by one will be treated as accepted by the other.
 - ssh-identity (default: off)
    Set to **on** to allow RadSSH to use ~/.ssh/id_rsa, ~/.ssh/id_dsa, and ~/.ssh/identity keys
 - ssh-agent (default: off)
    Set to **on** to allow RadSSH to connect to running ssh-agent/keyring service for keys.
 - authfile (default: ~/.radssh_auth)
    Optional authentication file, to store collections of passwords and keys to use on a per host basis.
 - plugins (default: None)
    Set to a comma separated list of directories where RadSSH should look for user plugins. The system level plugin directory inside the RadSSH package is always searched, regardless of this setting; this is for additional plugin directories to be specified.
 - disable_plugins (default: None)
    Set to a comma separated list of plugins to bypass loading.
 - quota.time (default: 0)
    Avoid runaway command execution by having RadSSH abort commands if host does not produce output for a given duration (in seconds). Setting of 0 = Unlimited.
 - quota.lines (default: 0)
    Avoid runaway command execution by having RadSSH abort commands if host produces too many lines of output. Setting of 0 = Unlimited.
 - quota.bytes (default: 0)
    Avoid runaway command execution by having RadSSH abort commands if host produces too many bytes of output. Setting of 0 = Unlimited.
 - commands.forbidden (default: telnet,ftp,sftp,vi,vim,ssh)
    Prevent use of the comma separated list of programs. Anything that needs interactive keyboard input will not likely behave as anticipated under RadSSH, and should not be run.
 - commands.restricted (default: rm,reboot,shutdown,halt,poweroff,telinit)
    Have RadSSH intercept possibly dangerous commands (extremely dangerous if mistakenly run on hundreds of servers simultaneously) and require explicit confirmation that the user intends to do precisely what was typed in.
 - paramiko_log_level
    Deprecated setting. Now able to be set via **loglevel** setting.
 - try_auth_none (default: off)
    Perform a initial authentication probing request to determine whether the remote host accepts keys or passwords, or both. Setting to **on** may improve connection speeds by bypassing unsupported authentication attempts, but use caution, as some remote SSH implementations, like Cisco switches will abruptly drop connection if auth-none is attempted.  OpenSSH on RHEL/CentOS 5 will fail to send a banner unless auth-none is attempted.
 - force_tty=Cisco,force10networks
    Set to a comma separated list of SSH host identifiers for connections that do not support SSH exec_command. This triggers a secondary, less reliable command invocation that runs commands through a dedicated tty session. Both Cisco and Force10 switches have been identified as requiring RadSSH operate in this mode; there may be others.
 - force_tty.signon (default: "term length 0")
    When a TTY session is required, RadSSH will issue this command after initial signon. For switches, this should avoid accumulating several "--More--" prompts in the output.
 - force_tty.signoff (default: "term length 20")
    When a TTY session is required, RadSSH will issue this command prior to a clean termination.


RELEASE_NOTES.md
Features
==========
 - New module for host key handling and validation, closer to the functionality of OpenSSH. Significantly improved performance when operating with very large known_hosts files.

 - Now relies on standard OpenSSH config file settings for options related to connection, host key validation, and user authentiction. Many, but not all options are supported.

    #### Connection Options Supported
    Hostname, Port, LogLevel, ConnectTimeout, Ciphers, KexAlgorithms, Macs, ProxyCommand, LocalCommand, PermitLocalCommand

    #### Hostkey Validation Options Supported
    GlobalKnownHostsFile, UserKnownHostsFile, StrictHostKeyChecking, HostKeyAlias, HostKeyAlgorithms, HashKnownHosts, CheckHostIP

    #### User Authentication Options Supported
    User, PreferredAuthentications, KbdInteractiveAuthentiction, PasswordAuthentication, PubkeyAuthentication, NumberOfPasswordPrompts, IdentityFile, IdentitiesOnly, BatchMode

 - Added support for ECDSA key files for user authentication.

 - Now supports for Paramiko version 2.0, and cryptography.io. Installations using Paramiko 1.X can continue to be used along with the unmaintained PyCrypto module.

  - Improved support for specifying explicit port and username with URI style formatting (user@host:port).

Issues Fixed
==========
 - A missing *plugins* directory is no longer a fatal runtime error.
 - StreamBuffer closing is no longer vulnerable to **Queue.Full** exception.
 - Fix thread contention issue when prompting user for passwords and accepting new host keys concurrently.
 - Ordered output mode fixed under Python3. [ Fixed in 1.1.1 ]
 - Improve behavior when connection is dropped by server during authentication. [ Fixed in 1.1.1 ]

Enhancements
============
 - Reworded user prompt when encountering a restricted command.
 - Command result summary explicitly lists return codes and hosts when any host returns a non-zero status.
 - Formerly deprecated configuration options are now categorized as "obsolete", since attempts to use them are ignored.
 - New "obsolete" configuration options, and their corresponding SSH Config option names:

    Obsolete RadSSH Option|Current SSH Option
    -------------------------|------------------------
    hostkey.verify|StrictHostKeyChecking
    hostkey.known_hosts|UserKnownHostsFile
    ssh_identity|IdentityFile
    ssh_agent|IdentitiesOnly

 - Added new config option **ssh_config** to allow overriding the default location of the user SSH Config file (~/.ssh/config).


Plugin Enhancements
=================
 - **\*enable** now includes explicit count of hosts when all hosts enabled.
 - **\*drop** with no arguments will now drop hosts that are not connected/authenticated.
 - **\*add** now handles URI style (user@host:port) format; lists summary of connections added to cluster.
 - Now supports shell alias definitions with escaped single quotes are supported.
 - **\*result** can include > and >> to save/append to local file [ Added in 1.1.1 ]
 - **\*history** added to alias plugin, with support for `!nnn` replay of command by history number [ Added in 1.1.1 ]
 - **\*tty** should no longer be prone to "Resource temporarily unavailable" exceptions. [ Fixed in 1.1.1 ]

API Changes
==========
 - hostkey module will be removed in 2.0 release in favor of the new known_hosts module.
 - AuthManager will no longer support parameters **include_agent** or **include_userkeys**. Settings are now controlled by the OpenSSH configuration options.

Known Issues
==========
 - Paramiko ProxyCommand may not function correctly under Python3 (https://github.com/paramiko/paramiko/issues/673)
 - ~~Ability to mix & match hostnames and IP addresses leads to many sorting issues under Python3 (#21)~~ [ Fixed in 1.1.1 ]

Additional Notes
==============
 - **PyCrypto** identified as an unmaintained module. **Paramiko** 2.0 has transitioned to using **cryptography**. RadSSH usage will prefer to use **cryptography**, but will continue to work with **PyCrypto** and **Paramiko** 1.X. Installations are strongly recommended to upgrade to using **cryptography** (and **Paramiko** 2.0) from **PyCrypto**. When RadSSH is unable to load the preferred **cryptography** module and reverts to using **PyCrypto**, a warning will be issued encouraging users to install **cryptography**.
 - **Paramiko** library does not include support for OpenSSH configuration *Match* blocks, so currently these are not supported for RadSSH configuration options as well.
 - Python 2.6 and Python 3.2 will continue to be supported by RadSSH 1.1.0, even though these Python versions are no longer being actively maintained. Future releases of RadSSH may not continue focus or effort on compatibility with these python versions.


radssh/config.py
#
# Copyright (c) 2014, 2016 LexisNexis Risk Data Management Inc.
#
# This file is part of the RadSSH software package.
#
# RadSSH is free software, released under the Revised BSD License.
# You are permitted to use, modify, and redsitribute this software
# according to the Revised BSD License, a copy of which should be
# included with the distribution as file LICENSE.txt
#

'''Configuration file module'''
from __future__ import print_function  # Requires Python 2.6 or higher


import sys
import os
import warnings
try:
    from io import StringIO
except ImportError:
    from StringIO import StringIO


system_settings_file = '/etc/radssh_config'
obsoleted = {
    'verbose': 'Use loglevel=[CRITICAL|ERROR|WARNING|INFO|DEBUG] instead',
    'paramiko_log_level': 'Use loglevel for controlling RadSSH and Paramiko logging',
    'hostkey.verify': 'Set StrictHostKeyChecking in standard SSH Config file (~/.ssh/config)',
    'hostkey.known_hosts': 'Set UserKnownHostsFile in standard SSH Config file (~/.ssh/config)',
    'ssh-identity': 'Set IdentityFile in standard SSH Config file (~/.ssh/config)',
    'ssh-agent': 'Set IdentitiesOnly in standard SSH Config file (~/.ssh/config)'
}

default_config = '''
# Package supplied configuration settings
### All lines starting with # are comments
### All other lines should be of the form keyword=value

### Shell Behavior Defaults
# NOTE: verbose and paramiko_log_level settings are obsolete
# Use loglevel instead, and logging from both RadSSH and paramiko get captured to radssh.log
#
# loglevel can be set to [CRITICAL|ERROR|WARNING|INFO|DEBUG]
loglevel=ERROR

shell.prompt=RadSSH $
shell.console=color
# To see last few lines of output of in-flight jobs when pressing Ctrl-C
# set this parameter to the number of lines to retain. Can be very memory
# intensive.
stalled_job_buffer=0

max_threads=120
# Automatically save log files into date/time-stamped local directory
logdir=session_%Y%m%d_%H%M%S
# Log all normal output to given filename in logdir. Set empty to turn off
# NOTE: This logging is in addition to host-by-host logging
log_out=out.log
# Log all error output to given filename in logdir. Set empty to turn off
# NOTE: This is in addition to host-by-host error logging
log_err=err.log

# Command line history file, saved across sessions
historyfile=~/.radssh_history

# Available modes: {stream, ordered, grouped, off}
output_mode=stream
# Can override character encoding (will use sys.stdout.encoding if not specified)
# character_encoding=UTF-8
# Avoiding runaway commands with either too much output, or
# waiting indefinately at a user prompt...
quota.time=0
quota.lines=0
quota.bytes=0

# Connection & Authentication Options
# Username defaults to $SSH_USER (or $USER) if not set here
# username=root
# Supplemental authentication file for more keys and/or passwords
authfile=~/.radssh_auth

# Network Tweaks
socket.timeout=30
keepalive=180

# Extensions to the shell via plugins
# System plugin collections always loaded from ${EXEC}/plugins
# User plugins loaded from directories listed here (optional)
plugins=
# Selectively disable plugins, if for some reason you want to
# avoid loading some system (or user) plugins that normally load
disable_plugins=

# Options hostkey.verify and hostkey.known_hosts are obsoleted
# in favor of using standard SSH Config settings StrictHostKeyChecking
# and UserKnownHostsFile. RadSSH can use a different config settings
# file by overriding the path to 'ssh_config' here:
ssh_config=~/.ssh/config

# Enable loading of user specific settings (and command line options)
# Only if this is set.
user.settings=~/.radssh_config

# Rudimentary command safeguards
# First, outright forbid commands that should only be run with a TTY
# which RadSSH typically does not provide...
commands.forbidden=telnet,ftp,sftp,vi,vim,ssh
# Also, for commands that could have devastating side effects, have
# RadSSH prompt the user if they are 100% sure they want to run...
commands.restricted=rm,reboot,shutdown,halt,poweroff,telinit

# Some SSH hosts do not support exec_command invocation, and require
# a session to be used via invoke_shell() instead. In addition, based
# on observation, the underlying transport is terminated when the session
# is closed, so the open session MUST stay persistent across command
# invocations.
# Identify such "problem" hosts by their SSH Server version string, which
# can be gathered using ssh-keyscan.
force_tty=Cisco,force10networks
# In addition, allow custom command(s) to be issued at signon and signoff
# to permit session to behave as RadSSH expects, without user input...
force_tty.signon=term length 0
force_tty.signoff=term length 20

# Should RadSSH initially send auth_none request (needed for OpenSSH 4.3 banner)
try_auth_none=off
'''

if sys.version_info[:2] == (2, 7):
    default_config = unicode(default_config)


def load_settings_file(f):
    '''Load settings from a file-like object, returning a dict'''
    settings = {}
    for line_number, line in enumerate(f, 1):
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        if '=' in line:
            setting = line.split('=', 1)
            settings[setting[0].strip()] = setting[1].strip()
        else:
            if hasattr(f, 'name'):
                warnings.warn_explicit('Invalid line in settings file [%s]' % line, RuntimeWarning, f.name, line_number)
            else:
                warnings.warn(RuntimeWarning('Invalid line in default settings (Line %d) "%s"' % (line_number, line)))
    return settings


def obsoleted_check(d, filename=None):
    '''Check settings dict against the obsoleted options'''
    for k in obsoleted:
        if k in d:
            warnings.warn('OBSOLETE: [%s] found in %s is ignored.\n\t%s' % (k, filename, obsoleted[k]))
            d.pop(k)


def load_default_settings():
    '''Load just the default settings, ignoring system and user settings files'''
    # Start with the default_config settings from the module
    defaults = StringIO(default_config)
    settings = load_settings_file(defaults)
    # Fill default username and character encodings from derived values
    # not in the default_config template string.
    if 'username' not in settings:
        settings['username'] = os.environ.get(
            'SSH_USER', os.environ.get('USER', os.environ.get('USERNAME', 'default')))
    if 'character_encoding' not in settings:
        if sys.stdout.encoding:
            settings['character_encoding'] = sys.stdout.encoding
        else:
            settings['character_encoding'] = 'UTF-8'
    return settings


def command_line_settings(cmdline_args=[], permitted=True):
    '''Load settings from command line args of the form "--keyword=value"'''
    settings = {}
    for arg in list(cmdline_args):
        if arg.startswith('--'):
            try:
                k, v = arg.split('=', 1)
                # Only apply command line option if user.settings is still enabled
                if permitted:
                    commandline_setting = {k[2:]: v}
                    obsoleted_check(commandline_setting, 'command line argument')
                    settings.update(commandline_setting)
                else:
                    warnings.warn(RuntimeWarning('Command line option: %s (ignored) - User settings disabled by administrator' % (arg)))
            except ValueError:
                warnings.warn(RuntimeWarning('Invalid command line option: %s (ignored)' % (arg)))
            cmdline_args.remove(arg)
    return settings


def load_settings(cmdline_args=None):
    '''Load a full settings dict from defaults, system, user, and command line settings'''
    settings = load_default_settings()
    # Also load the system-wide settings
    if os.path.exists(system_settings_file):
        with open(system_settings_file) as f:
            system_settings = load_settings_file(f)
            obsoleted_check(system_settings, system_settings_file)
            settings.update(system_settings)
    # If admin has not disabled user settings, load them and finally the command line settings
    if settings.get('user.settings'):
        user_settings_file = os.path.expanduser(settings.get('user.settings'))
        if os.path.exists(user_settings_file):
            with open(user_settings_file) as f:
                user_settings = load_settings_file(f)
                obsoleted_check(user_settings, user_settings_file)
                settings.update(user_settings)
    if cmdline_args and settings.get('user.settings'):
        if isinstance(cmdline_args, list):
            cmdline_dict = command_line_settings(cmdline_args)
            settings.update(cmdline_dict)
        else:
            settings.update(cmdline_args)
    return settings


def main():
    '''Print the RadSSH default settings as reference'''
    for x in default_config.split('\n'):
        if x.startswith('#'):
            print('##%s' % str(x))
        elif not x.strip():
            print()
        else:
            print('# %s' % str(x))

if __name__ == '__main__':
    main()
